libdrgn/linux_kernel_helpers.c

```diff
@@ -556,380 +536,82 @@ struct drgn_error *linux_helper_find_task(struct drgn_object *res,
-	if (!frame)
-		return &drgn_enomem;
-	frame->index = index;
-	frame->next_slot = 0;
-	drgn_object_init(&frame->slots, iter->prog);
-	// We temporarily use `frame->slots` to hold `shift` in order to avoid
-	// using another `struct drgn_object`.
-	err = drgn_object_member_dereference(&frame->slots, node, "shift");
-	if (err)
-		goto err_frame;
-	err = drgn_object_read_unsigned(&frame->slots, &frame->shift);
-	if (err)
-		goto err_frame;
-	// Now `frame->slots` is actually used for `slots`.
-	err = drgn_object_member_dereference(&frame->slots, node, "slots");
-	if (err)
-		goto err_frame;
-	return NULL;
-
-err_frame:
-	drgn_object_deinit(&frame->slots);
-	linux_helper_radix_tree_iter_frame_vector_pop(&iter->frames);
-	return err;
-}
-
-struct drgn_error *linux_helper_radix_tree_iter_next(struct linux_helper_radix_tree_iter *iter,
-						     struct linux_helper_radix_tree_iter_entry **ret)
-{
-	bool entry_populated = false;
-	struct drgn_error *err = NULL;
-	struct drgn_object node;
-	drgn_object_init(&node, iter->prog);
-	if (!iter->started) {
-		iter->started = true;
-		err = radix_tree_iter_handle_node(iter, &iter->root, 0, &entry_populated);
-	}
-
-	while (!err && !entry_populated && iter->frames.size) {
-		struct linux_helper_radix_tree_iter_frame *frame =
-			&iter->frames.data[iter->frames.size - 1];
-		if (frame->next_slot <= iter->RADIX_TREE_MAP_MASK) {
-			err = drgn_object_subscript(&node, &frame->slots, frame->next_slot);
-			if (!err)
-				err = radix_tree_iter_handle_node(iter, &node,
-								  frame->index + (frame->next_slot++
-										  << frame->shift),
-								  &entry_populated);
-		} else {
-			drgn_object_deinit(&frame->slots);
-			linux_helper_radix_tree_iter_frame_vector_pop(&iter->frames);
-		}
-	}
-	if (!err)
-		*ret = entry_populated ? &iter->entry : NULL;
-	drgn_object_deinit(&node);
-	return err;
-}
-
-struct drgn_error *linux_helper_idr_iter_init(struct linux_helper_idr_iter *iter,
-					      const struct drgn_object *idr)
-{
-	struct drgn_error *err;
-	struct drgn_object idr_rt, idr_base;
-	drgn_object_init(&idr_rt, drgn_object_program(idr));
-	drgn_object_init(&idr_base, drgn_object_program(idr));
-
-	err = drgn_object_member(&idr_base, idr, "idr_base");
-	if (!err) {
-		err = drgn_object_read_unsigned(&idr_base, &iter->base);
-		if (err)
-			goto out;
-	} else if (err->code == DRGN_ERROR_LOOKUP) {
-		drgn_error_destroy(err);
-		iter->base = 0;
-	} else {
 		goto out;
-	}
-
-	err = drgn_object_member(&idr_rt, idr, "idr_rt");
+	err = drgn_object_address_of(&iter->task, &iter->task);
 	if (err)
 		goto out;
-	err = drgn_object_address_of(&idr_rt, &idr_rt);
+	err = drgn_object_read_unsigned(&iter->task, &iter->init_task_address);
 	if (err)
 		goto out;
-	err = linux_helper_radix_tree_iter_init(&iter->iter, &idr_rt);
+	iter->thread_group_address = iter->init_task_address;
+
 out:
-	drgn_object_deinit(&idr_rt);
-	drgn_object_deinit(&idr_base);
+	if (err)
+		drgn_object_deinit(&iter->task);
 	return err;
```

Comment:
Style nit: since the successful case and the error case don't share any cleanup, I think it makes more sense to do:

```c
	err = ...;
	if (err)
		goto err;
	...
	return NULL;

err:
	drgn_object_deinit(&iter->task);
	return err;
```